### PR TITLE
Implement constant propagation on top of MIR SSA analysis

### PR DESCRIPTION
**Pull request by cjgillot**
This implements the idea I proposed in https://github.com/rust-lang/rust/pull/110719#issuecomment-1718324700

Based on https://github.com/rust-lang/rust/pull/109597

The value numbering "GVN" pass formulates each rvalue that appears in MIR with an abstract form (the `Value` enum), and assigns an integer `VnIndex` to each. This abstract form can be used to deduplicate values, reusing an earlier local that holds the same value instead of recomputing. This part is proposed in #109597.

From this abstract representation, we can perform more involved simplifications, for example in https://github.com/rust-lang/rust/pull/111344.

With the abstract representation `Value`, we can also attempt to evaluate each to a constant using the interpreter. This builds a `VnIndex -> OpTy` map. From this map, we can opportunistically replace an operand or a rvalue with a constant if their value has an associated `OpTy`.

The most relevant commit is [Evaluated computed values to constants.](https://github.com/rust-lang/rust/commit/2767c4912ea249c2f613a9cedcd6c13ea1237e54)"

r? @oli-obk

### CODE REVIEW

File: tests/codegen/issues/issue-105386-ub-in-debuginfo.rs [hunk 1]
```diff
@@ -19,5 +19,3 @@ pub fn outer_function(x: S, y: S) -> usize {
 // CHECK-NOT: [[ptr_tmp:%.*]] = getelementptr inbounds %"{closure@{{.*.rs}}:9:23: 9:25}", ptr [[spill]]
 // CHECK-NOT: [[load:%.*]] = load ptr, ptr
 // CHECK: call void @llvm.lifetime.start{{.*}}({{.*}}, ptr [[spill]])
-// CHECK: [[inner:%.*]] = getelementptr inbounds %"{{.*}}", ptr [[spill]]
-// CHECK: call void @llvm.memcpy{{.*}}(ptr {{align .*}} [[inner]], ptr {{align .*}} %x
```
Comment (oli-obk) [inline tests/codegen/issues/issue-105386-ub-in-debuginfo.rs:23]: should this be changed to `CHECK-NOT` so we know this doesn't come back?

**Review by oli-obk — APPROVED** [review]
Looks like we lost `Len`, `BitAnd`/`BitOr` and some field access/debug info optimizations, but those can be readded later

**Comment by oli-obk** [comment]
r=me with the codegen test adjusted to prevent silent future regressions.

**Comment by cjgillot** [comment]
I'll also wait for some progress on https://github.com/rust-lang/rust/issues/116971 to avoid unseen regressions in const_prop.

**Comment by Dylan-DPC** [comment]
Marking this as blocked on #116971

**Comment by cjgillot** [comment]
Blocked on https://github.com/rust-lang/rust/issues/117761

**Comment by cjgillot** [comment]
https://github.com/rust-lang/rust/issues/117761 is merged. Beta has branched off, so this will have time to cook.
Per https://github.com/rust-lang/rust/pull/116012#issuecomment-1787052524,
@bors r=oli-obk

**Comment by cjgillot** [comment]
@saethlin took a look a the last commit. 
@bors r=oli-obk

File: compiler/rustc_infer/src/infer/relate/generalize.rs [hunk 2]
```diff
@@ -226,7 +227,9 @@ where
         let old_ambient_variance = self.ambient_variance;
         self.ambient_variance = self.ambient_variance.xform(variance);
         debug!(?self.ambient_variance, "new ambient variance");
-        let r = self.relate(a, b)?;
+        // Recursive calls to `relate` can overflow the stack. For example a deeper version of
```
[the remark this answers is not in the document]
Comment (cjgillot) [reply compiler/rustc_infer/src/infer/relate/generalize.rs:230]: This is a really deep type.

I'm a bit guessing what the stack trace from CI points to. The failing test uses a deep type, and by making it even deeper I got a backtrace pointing involving this recursion.

[the remark this answers is not in the document]
Comment (compiler-errors) [reply compiler/rustc_infer/src/infer/relate/generalize.rs:230]: Discussed this offline. I guess this `ensure_sufficient_stack` invocation is fine.

**Comment by saethlin** [comment]
Looks to me like the timeout was in uploading to s3

@bors retry

**Comment by rylev** [comment]
It seems given the positives for this change and the fact that improvements outweigh regressions, we can call this triaged.

@rustbot label: +perf-regression-triaged

**Comment by RalfJung** [comment]
I'm only just seeing this... this is amazing, old const-prop is gone. :) Thanks a ton @cjgillot!

**Pull request merged**